File: apps/analyzer/app/observability/metrics.py
from __future__ import annotations

import os
import time
from typing import Any, Literal

from prometheus_client import CONTENT_TYPE_LATEST, Counter, Histogram, generate_latest, start_http_server

AiRequestStatus = Literal["success", "error"]
AiRequestErrorType = Literal[
    "none",
    "disabled",
    "missing_api_key",
    "timeout",
    "http_error",
    "invalid_json",
    "network_error",
    "unknown",
]

AI_REQUESTS_TOTAL = Counter(
    "wedge_ai_gms_requests_total",
    "Total number of Wedge AI GMS requests by service, feature, model, status, and error type.",
    ["service", "feature", "model", "status", "error_type"],
)

AI_REQUEST_DURATION_SECONDS = Histogram(
    "wedge_ai_gms_request_duration_seconds",
    "Duration of Wedge AI GMS requests in seconds by service, feature, model, status, and error type.",
    ["service", "feature", "model", "status", "error_type"],
    buckets=(0.1, 0.25, 0.5, 1, 2.5, 5, 10, 20, 30),
)


def include_metrics_route(app: Any) -> None:
    if not _env_bool("ANALYZER_METRICS_ENABLED", default=False):
        return

    from fastapi import Response

    @app.get("/metrics", include_in_schema=False)
    def metrics() -> Response:
        return Response(generate_latest(), media_type=CONTENT_TYPE_LATEST)


def start_metrics_server_from_env() -> None:
    if not _env_bool("ANALYZER_METRICS_ENABLED", default=False):
        return
    host = os.environ.get("ANALYZER_METRICS_HOST", "0.0.0.0")
    port = _env_int("ANALYZER_METRICS_PORT", default=9102)
    start_http_server(port=port, addr=host)
    print(f"analyzer metrics server started: host={host} port={port}", flush=True)


def observe_gms_request(
    *,
    feature: str,
    model: str,
    status: AiRequestStatus,
    error_type: AiRequestErrorType,
    started_at: float,
) -> None:
    labels = {
        "service": "analyzer",
        "feature": _label_value(feature, "unknown"),
        "model": _label_value(model, "unknown"),
        "status": status,
        "error_type": error_type,
    }
    AI_REQUESTS_TOTAL.labels(**labels).inc()
    AI_REQUEST_DURATION_SECONDS.labels(**labels).observe(max(0.0, time.perf_counter() - started_at))


def _label_value(value: str, fallback: str) -> str:
    stripped = value.strip()
    return stripped if stripped else fallback


def _env_bool(name: str, *, default: bool) -> bool:
    value = os.environ.get(name)
    if value is None:
        return default
    return value.strip().lower() in {"1", "true", "yes", "on"}


def _env_int(name: str, *, default: int) -> int:
    value = os.environ.get(name)
    if value is None:
        return default
    try:
        parsed = int(value)
    except ValueError:
        return default
    return parsed if parsed > 0 else default

File: apps/runner/src/observability/metrics.ts
import { createServer, type Server } from "node:http";
import { Counter, Histogram, register } from "prom-client";
import type { RunnerConfig } from "../config/index.ts";
import { errorMessage, logOperationalEvent } from "../shared/utils.ts";

const AI_REQUESTS_TOTAL_NAME = "wedge_ai_gms_requests_total";
const AI_REQUEST_DURATION_SECONDS_NAME = "wedge_ai_gms_request_duration_seconds";

type AiRequestStatus = "success" | "error";
export type AiRequestErrorType =
  | "none"
  | "timeout"
  | "http_error"
  | "invalid_json"
  | "network_error"
  | "unknown";

export interface AiRequestMetricsInput {
  service: string;
  feature: "agent_decision" | "scenario_authoring";
  model: string;
  status: AiRequestStatus;
  errorType: AiRequestErrorType;
  durationMs: number;
}

export interface RunnerMetricsServer {
  close: () => Promise<void>;
}

const aiRequestsTotal = new Counter({
  name: AI_REQUESTS_TOTAL_NAME,
  help: "Total number of Wedge AI GMS requests by service, feature, model, status, and error type.",
  labelNames: ["service", "feature", "model", "status", "error_type"] as const
});

const aiRequestDurationSeconds = new Histogram({
  name: AI_REQUEST_DURATION_SECONDS_NAME,
  help: "Duration of Wedge AI GMS requests in seconds by service, feature, model, status, and error type.",
  labelNames: ["service", "feature", "model", "status", "error_type"] as const,
  buckets: [0.1, 0.25, 0.5, 1, 2.5, 5, 10, 20, 30]
});

export function recordAiRequestMetrics(input: AiRequestMetricsInput): void {
  const labels = {
    service: input.service,
    feature: input.feature,
    model: normalizeMetricLabel(input.model, "unknown"),
    status: input.status,
    error_type: input.errorType
  };

  aiRequestsTotal.inc(labels);
  aiRequestDurationSeconds.observe(labels, Math.max(0, input.durationMs) / 1000);
}

export function startRunnerMetricsServer(config: RunnerConfig): RunnerMetricsServer | null {
  if (!config.metricsEnabled) {
    return null;
  }

  const server = createServer(async (request, response) => {
    if (request.url !== "/metrics") {
      response.writeHead(404, { "content-type": "text/plain; charset=utf-8" });
      response.end("not found\n");
      return;
    }

    try {
      response.writeHead(200, { "content-type": register.contentType });
      response.end(await register.metrics());
    } catch (error) {
      response.writeHead(500, { "content-type": "text/plain; charset=utf-8" });
      response.end(`${errorMessage(error)}\n`);
    }
  });

  server.listen(config.metricsPort, config.metricsHost, () => {
    logOperationalEvent("runner-metrics", "server_started", {
      host: config.metricsHost,
      port: config.metricsPort,
      path: "/metrics"
    });
  });

  server.on("error", (error) => {
    logOperationalEvent("runner-metrics", "server_error", {
      reason: errorMessage(error)
    }, "error");
  });

  return {
    close: () => closeServer(server)
  };
}

function normalizeMetricLabel(value: string, fallback: string): string {
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : fallback;
}

async function closeServer(server: Server): Promise<void> {
  await new Promise<void>((resolve, reject) => {
    server.close((error) => {
      if (error) {
        reject(error);
        return;
      }
      resolve();
    });
  });
}
